File: v1/java/custom-set/src/main/java/CustomSet.java
import java.util.*;

public class CustomSet<T> {
    private final Map<Integer, T> elements;

    public CustomSet(List<T> input) {
        elements = new HashMap<>();

        for (T element: input) {
            add(element);
        }
    }

    public void add(T element) {
        if (contains(element))
            return;

        elements.put(element.hashCode(), element);
    }

    public boolean isEmpty() {
        return elements.isEmpty();
    }

    public boolean contains(T value) {
        return elements.containsKey(value.hashCode());
    }

    public boolean isSubset(CustomSet<T> other) {
        for(Map.Entry<Integer, T> entry : other.elements.entrySet()) {
            if (!contains(entry.getValue()))
                return false;
        }

        return true;
    }

    public boolean isDisjoint(CustomSet<T> other) {
        for(Map.Entry<Integer, T> entry : other.elements.entrySet()) {
            if (contains(entry.getValue()))
                return false;
        }

        return true;
    }

    public CustomSet<T> getIntersection(CustomSet<T> other) {
        List<T> values = new LinkedList<>();

        elements.forEach((k, v) -> {
            if (other.contains(v))
                values.add(v);
        });

        return new CustomSet<T>(values);
    }

    public CustomSet<T> getDifference(CustomSet<T> other) {
        List<T> values = new LinkedList<>();

        elements.forEach((k, v) -> {
            if (!other.contains(v))
                values.add(v);
        });

        return new CustomSet<T>(values);
    }

    public CustomSet<T> getUnion(CustomSet<T> other) {
        List<T> values = new LinkedList<>(elements.values());
        values.addAll(other.elements.values());

        return new CustomSet<T>(values);
    }

    public boolean equals(Object other){
        if (other instanceof CustomSet<?>){
            if (((CustomSet<?>)other).elements.keySet().equals(elements.keySet())){
                return true;
            }
        }

        return false;
    }
}

File: kotlin/custom-set/src/main/kotlin/CustomSet.kt
class CustomSet(vararg elements: Int) {
    private val hashToElement = elements.associateBy { it.hashCode() }.toMutableMap()

    fun isEmpty(): Boolean = hashToElement.isEmpty()

    fun isSubset(other: CustomSet) = hashToElement.keys.all { it in other.hashToElement }

    fun isDisjoint(other: CustomSet) = hashToElement.keys.all { it !in other.hashToElement }

    fun contains(other: Int) = other.hashCode() in hashToElement

    fun add(other: Int) {
        hashToElement.putIfAbsent(other.hashCode(), other)
    }

    override fun equals(other: Any?) = when(other) {
        is CustomSet -> other.isSubset(this) && this.isSubset(other)
        else -> false
    }

    fun intersection(other: CustomSet) = CustomSet().also { newSet ->
        hashToElement.filter { it.key in other.hashToElement }.values.forEach { newSet.add(it) } }

    operator fun plus(other: CustomSet) = CustomSet().also { newSet ->
        hashToElement.values.forEach { newSet.add(it) }
        other.hashToElement.values.forEach { newSet.add(it) }
    }

    operator fun minus(other: CustomSet) = CustomSet().also { newSet ->
        hashToElement.filter { it.key !in other.hashToElement }.values.forEach { newSet.add(it) }
    }
}
